lib/rex/proto/rfb/client.rb

```diff
@@ -14,243 +14,242 @@
-    if not result
-      @error = "Unable to read auth result"
-      return false
-    end
-
-    result = result.unpack('N').first
-    case result
-    when 0
-      return true
-
-    when 1
-      if @minver >= 8
-        msg = read_error_message
-        @error = "Authentication failed: #{msg}"
-      else
-        @error = "Authentication failed"
-      end
-    when 2
-      @error = "Too many authentication attempts"
-    else
-      @error = "Unknown authentication result: #{result}"
-    end
-
-    false
-  end
-
-  def negotiate_authentication
-    # Authentication type negotiation is protocol version specific.
-    if @minver < 7
-      buf = @sock.get_once(4)
-      if not buf
-        @error = "Unable to obtain requested authentication method"
-        return nil
-      end
-      @auth_types = buf.unpack('N')
-      if not @auth_types or @auth_types.first == 0
-        msg = read_error_message
-        @error = "No authentication types available: #{msg}"
-        return nil
-      end
-    else
-      buf = @sock.get_once(1)
-      if not buf
-        @error = "Unable to obtain supported authentication method count"
-        return nil
-      end
-
-      # first byte is number of security types
-      num_types = buf.unpack("C").first
-      if (num_types == 0)
-        msg = read_error_message
-        @error = "No authentication types available: #{msg}"
-        return nil
+  module Proto
+    module RFB
+      class Client
+        include Rex::Proto::RFB::Constants
+        def initialize(sock, opts = {})
+          @sock = sock
+          @opts = opts
+
+          @banner = nil
+          @majver = -1
+          @minver = -1
+          @auth_types = []
+        end
+
+        def read_error_message
+          len = @sock.get_once(4)
+          return 'Unknown error' if !len or len.length != 4
+
+          len = len.unpack('N').first
+          @sock.get_once(len)
+        end
+
+        def handshake
+          @banner = @sock.get_once(12)
+          if !@banner
+            @error = 'Unable to obtain banner from server'
+            return false
+          end
+
+          # RFB Protocol Version 3.3 (1998-01)
+          # RFB Protocol Version 3.7 (2003-08)
+          # RFB Protocol Version 3.8 (2007-06)
+
+          if @banner =~ /RFB ([0-9]{3})\.([0-9]{3})/
+            @majver = Regexp.last_match(1).to_i
+            unless Constants::MajorVersions.include?(@majver)
+              @error = "Invalid major version number: #{@majver}"
+              return false
+            end
+          else
+            @error = "Invalid RFB banner: #{@banner}"
+            return false
+          end
+
+          @minver = Regexp.last_match(2).to_i
+
+          # Forces version 3 to be used. This adds support for version 4 servers.
+          # Minor version is at least 3, to work with Apple devices. Attempting to be purely dynamic
+          # based on the received server version can result in a 5.0 setting the client 3.0
```

Comment:
```suggestion
          # Minor version is at least 3, to work with Apple devices. Attempting to be purely dynamic
          # based on the received server version can result in a 5.0 server version response setting
          # our client version to 3.0 when it should be at least 3.3
```

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/16207/commits/4e8613904ae32b9744bb8eb79afc851f341d2d14